### PR TITLE
Use HostingNullLogScope when no logging

### LINKED ISSUE #367: [Perf] High Allocator LoggingScope when no loggers
Using works with null references, so this is simpler than outlined below; added change to logging to improve efficiency.

---
Creating a scope is expensive (at benchmarking speeds) and the techempower benchmarks specify in [general requirements](https://www.techempower.com/benchmarks/#section=code) item 2:

> All test implementations **must disable all disk logging**. For many reasons, we expect all tests will run without writing logs to disk. Most importantly, the volume of requests is sufficiently high to fill up disks even with only a single line written to disk per request. Please disable all forms of disk logging. We recommend but do not require disabling console logging as well.

For fair comparison, only create the logging scope if logging is enabled:

Add ```LogLevel``` of ```None = 7```? 

Change the following from:

```csharp
var logger = _applicationServices.GetRequiredService<ILogger<HostingEngine>>();
...
var server = ServerFactory.Start(_serverInstance,
    async features =>
    {
        try
        {
            using (logger.BeginScope("Request Id: {RequestId}", requestIdentifier))
            {
                contextAccessor.HttpContext = httpContext;
                await application(httpContext);
            }
        }
        catch (Exception ex)
        {
            if (telemetrySource.IsEnabled("Microsoft.AspNet.Hosting.UnhandledException"))
            {
                telemetrySource.WriteTelemetry("Microsoft.AspNet.Hosting.UnhandledException", new { httpContext = httpContext, exception = ex });
            }

            throw;
        }
```

To something similar to:

```csharp
var logger = _applicationServices.GetRequiredService<ILogger<HostingEngine>>();
...
var isLogging = logger.IsEnabled(LogLevel.Critical); // ? not sure about how not logging is checked
var server = ServerFactory.Start(_serverInstance,
    async features =>
    {
        ILogger loggerScope;
        try
        {
            if (isLogging) 
            {
                loggerScope = logger.BeginScope("Request Id: {RequestId}", requestIdentifier);
            }

            contextAccessor.HttpContext = httpContext;
            await application(httpContext);
        }
        catch (Exception ex)
        {
            if (telemetrySource.IsEnabled("Microsoft.AspNet.Hosting.UnhandledException"))
            {
                telemetrySource.WriteTelemetry("Microsoft.AspNet.Hosting.UnhandledException", new { httpContext = httpContext, exception = ex });
            }

            throw;
        }
        finally
        {
            if (loggerScope != null) loggerScope.Dispose();
        }
```

Ensure having no local logging scope doesn't cause issues.

### PR DESCRIPTION
**Pull request by benaadams**
Drops the `HostingLogScope` from this 
After-QueueUserWorkItemCallback

Resolves #367

/cc @davidfowl

### CODE REVIEW

**Comment by davidfowl** [comment]
Checking IsEnabled for the Hosting logger isn't enough. You need to be able to check if there are any logging providers at all and there's no way to do that.

/cc @lodejard @rynowak

**Comment by benaadams** [comment]
DI or lambda \ `Func<object, object>` scope?

e.g.
```csharp
return logger.BeginScopeImpl((state) => new HostingLogScope((HttpContext)state), httpContext);
```

File: src/Microsoft.AspNetCore.Hosting/Internal/HostingLoggerExtensions.cs [hunk 1]
```diff
@@ -306,6 +307,78 @@ IEnumerator IEnumerable.GetEnumerator()
...
+
+                object IEnumerator.Current
+                {
+                    get
+                    {
+                        return null;
+                    }
+                }
+
+                public void Dispose()
```
Comment (SajayAntony) [inline src/Microsoft.AspNetCore.Hosting/Internal/HostingLoggerExtensions.cs]: Elide?

**Comment by benaadams** [comment]
Need to see how this goes https://github.com/aspnet/Logging/issues/366

**Comment by davidfowl** [comment]
Closing until the logging issue is resolved.

**Pull request closed without merging**